### PR TITLE
[NOSQUASH] Script api fixes and preparations for `mglua`

### PR DESCRIPTION
**Pull request by sfan5**
independent commits split from my other PR so they can already be reviewed

## To do

This PR is Ready for Review.

### CODE REVIEW

File: src/script/lua_api/l_env.cpp [hunk 1]
```diff
@@ -991,16 +971,28 @@ int ModApiEnvMod::l_find_nodes_in_area_under_air(lua_State *L)
 	std::vector<content_t> filter;
 	collectNodeIds(L, 3, ndef, filter);
 
+	bool grouped = lua_isboolean(L, 4) && readParam<bool>(L, 4);
+
+	auto iterate = [&] (auto callback) {
```
Comment (Desour) [inline src/script/lua_api/l_env.cpp:976]: No need to copy.
```suggestion
	auto iterate = [&] (auto &&callback) {
```

**Review by Desour — COMMENTED** [review]
Fine otherwise.

The functional change of `set_node_level` now doing a swap (e.g. no `on_construct`/`on_destruct`) should *probably* be fine.

File: src/script/lua_api/l_env.cpp (hunk 1 above)
Comment (SmallJoker) [reply src/script/lua_api/l_env.cpp:976]: @Desour Not copy what? It's a function pointer.

File: src/script/scripting_server.cpp [hunk 2]
```diff
@@ -89,28 +89,29 @@ ServerScripting::ServerScripting(Server* server):
...
-		lua_getglobal(L, "core");
-		luaL_checktype(L, -1, LUA_TTABLE);
-		lua_getfield(L, -1, "get_globals_to_transfer");
-		lua_call(L, 0, 1);
-		auto *data = script_pack(L, -1);
-		assert(!data->contains_userdata);
-		getServer()->m_async_globals_data.reset(data);
-		lua_pushnil(L);
-		lua_setfield(L, -3, "get_globals_to_transfer"); // unset function too
-		lua_pop(L, 2); // pop 'core', return value
```
Comment (SmallJoker) [inline src/script/scripting_server.cpp:106]: This was not moved to the new code. Which function does pop `core` now?

File: src/script/lua_api/l_env.cpp (hunk 1 above)
Comment (Desour) [reply src/script/lua_api/l_env.cpp:976]: It's a lambda.
Not sure how bad the copying is if it's just capturing by reference, but still, there's no need to take the lambda by value.

File: src/script/lua_api/l_vmanip.cpp [hunk 3]
```diff
@@ -111,12 +111,14 @@ int LuaVoxelManip::l_set_data(lua_State *L)
...
 {
-	MAP_LOCK_REQUIRED;
+	GET_ENV_PTR;
 
 	LuaVoxelManip *o = checkObject<LuaVoxelManip>(L, 1);
 	bool update_light = !lua_isboolean(L, 2) || readParam<bool>(L, 2);
 
-	GET_ENV_PTR;
+	if (o->vm->isOrphan())
+		return 0;
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/script/lua_api/l_vmanip.cpp:120]: Probably not.

File: builtin/common/register.lua
[the remark this answers is not in the document]
Comment (sfan5) [reply builtin/common/register.lua:1]: IIRC the client code has some additional restrictions about debug.getinfo and also CSM will be totally reworked at some point anyway, so I opted not to refactor that too.

File: src/script/scripting_server.cpp [hunk 4]
```diff
@@ -89,23 +89,24 @@ ServerScripting::ServerScripting(Server* server):
...
-	}
+	SCRIPTAPI_PRECHECKHEADER
+
+	lua_getglobal(L, "core");
+	luaL_checktype(L, -1, LUA_TTABLE);
+	lua_getfield(L, -1, "get_globals_to_transfer");
+	lua_call(L, 0, 1);
+	auto *data = script_pack(L, -1);
+	assert(!data->contains_userdata);
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/script/scripting_server.cpp:101]: indeed, but they aren't supposed to
plus builtin already takes care to remove such objects from `registered_items`

File: src/script/scripting_server.cpp (hunk 2 above)
Comment (sfan5) [reply src/script/scripting_server.cpp:106]: `top` points to the core table and any handling of that is done by the caller somewhere else

File: src/script/lua_api/l_env.cpp (hunk 1 above)
Comment (sfan5) [reply src/script/lua_api/l_env.cpp:976]: FYI: I didn't opt for the potentially simpler (and type-checked) solution using `std::function` due to performance concerns

File: src/script/scripting_server.cpp (hunk 2 above)
Comment (Desour) [reply src/script/scripting_server.cpp:106]: An explicit pop would be cleaner imo.

File: src/script/scripting_server.cpp [hunk 5]
```diff
@@ -89,28 +89,30 @@ ServerScripting::ServerScripting(Server* server):
...
+	getServer()->m_lua_globals_data.reset(data);
+	// unset the function
+	lua_pushnil(L);
+	lua_setfield(L, -3, "get_globals_to_transfer");
+	lua_pop(L, 2); // pop 'core', return value
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/script/scripting_server.cpp:106]: from `lua_call(L, 0, 1);`

[the remark this answers is not in the document]
Comment (Desour) [reply src/script/scripting_server.cpp:106]: Ahh ok, "return-value" not "and do return the value".

**Review by Desour — APPROVED** [review]

**Pull request merged**